src/inet/TCPEndPoint.cpp

```diff
@@ -1844,7 +1844,10 @@ void TCPEndPoint::HandleDataSent(uint16_t lenSent)
     if (IsConnected())
     {
         // Consume data off the head of the send queue equal to the amount of data being acknowledged.
-        mSendQueue = mSendQueue->Consume(lenSent);
+        if (mSendQueue)
```

Comment:
I'm trying to work through the logic of the larger state machine to check whether it is possible for `mSendQueue` to be `NULL` at this point in time.  I note that `HandleDataSent` is private and thus should not be invoked at an arbitrary point in the execution.  It is processed as an event , and only as a result of the LwIP `tcp_sent` callback.  There is generally a benefit to having the system fail fast when the core invariants are violated; this change perhaps makes the system fail more gracefully at a cost of sweeping some issues under the rug.  If you have an identified codepath that leads to the check being required here, it would be helpful.

Reply:
I'm pretty sure this change is hiding a bigger problem.  As I see it, the scenario is: LwIP has called back into the InetLayer (via a call to LwIPHandleDataSent()) stating that a particular amount of previously sent data has been ACKed by the peer.  This caused an event to be posted to the Weave thread.  InetLayer::HandleInetLayerEvent() dispatched this event to TCPEndPoint::HandleDataSent().  At this point, the endpoint is still connected (IsConnected() == true).  However, the pointer to the buffer containing the sent data (mSendQueue) is now NULL.  

This state is inconsistent with the logic of TCPEndPoint, and of the LwIP API.  When using the LwIP raw API, the sender is required to preserve the buffer containing the data to be sent until it receives a callback indicating that all the data has been acknowledged by the peer. 

Now, PacketBuffer::Consume() will return a NULL if the given length consumes the entirety of the data in the buffer.  And, indeed, this is normal way that mSendQueue becomes NULL: the peer ACKs the data, Comsume() consumes the number of bytes ACKed, and if this results in the buffer being empty, it frees the buffer and returns NULL.

In theory, if LwIP generated an _extraneous_ callback to the data acknowledged function after all sent data had been previously acknowledged, then that would result in HandleDataSent() calling Consume() with a NULL this pointer.  (Note that if this did happen, then I would expect lenSent to be 0... because what else could it be?)  Perhaps this situation can arise when ACKs are retransmitted at the TCP level.  Although I still would consider this a bug in LwIP.

That theory aside, if mSendQueue is NULL, something more subtle is going on that needs to be understood.

Reply:
@jaylogue -- agreed.